File: BC - w4/java/Unit/Unit.java
package Unit;



public class Unit {
	
	private int damage;
	private int hitPoints;
	private int hitPointsLimit;
	private String name;
	
	
	public Unit(String name, int hp, int dmg){
		this.damage = dmg;
		this.hitPointsLimit = hp;
		this.name = name;
	    this.hitPoints = hitPointsLimit;
	    }

	private void ensureIsAlive() throws UnitIsDead{
	       if ( hitPoints == 0 ) {
	           throw new UnitIsDead ();
	       }
	    }

	public int getDamage() {
	        return this.damage;
	    }
	public int getHitPoints() {
	        return this.hitPoints;
	    }
	public int getHitPointsLimit() {
	        return this.hitPointsLimit;
	    }
	    String getName(){
	        return this.name;
	    }

	    public void addHitPoints(int hp) {
	        try {
	        ensureIsAlive();
	        int currentHitPoint = this.hitPoints;

	        currentHitPoint += hp;

	        if (currentHitPoint > hitPointsLimit) {
	            this.hitPoints = hitPointsLimit;
	        } else {
	            this.hitPoints =  currentHitPoint;
	        }
	        } catch ( Exception UnitIsDead) {
	        	System.out.println( "Cant Heal dead unit");
	        }


	    }
	    public void takeDamage(int dmg) {
	        int currentHitPoint = this.hitPoints;

	        currentHitPoint -= dmg;
	        if (currentHitPoint <= 0) {
	            currentHitPoint = 0;
	        } 
	        this.hitPoints = currentHitPoint;

	    }

	    public void attack(Unit enemy) {

	        try {
	        ensureIsAlive();
	        enemy.takeDamage(this.damage);
	        enemy.counterAttack(this);
	        } catch (Exception UnitIsDead) {
	        	System.out.println ("my target is dead");
	        }
	    }
	    public void counterAttack(Unit enemy) {
	        try {
	        ensureIsAlive();
	        this.takeDamage(enemy.damage/2);
	        } catch (Exception UnitIsDead) {
	        	System.out.println ("this unit is dead");
	        }
	    }

	    public void print(){
			System.out.println("unit "+this.name+"\n"
								+"damage " + this.damage+"\n"
								+"hit points " + this.hitPoints+ "\n"
								+"Max HP " + this.hitPointsLimit);
		}

}

File: BC - w4/Unit/unit.cpp
    #include "unit.h"


    Unit::Unit(const std::string& name, int hp, int dmg) :
    name(name),
    hitPointsLimit(hp),
    damage(dmg)
    {
        this->hitPoints = hitPointsLimit;
    }
    Unit::~Unit(){}

    void Unit::ensureIsAlive() {
       if ( hitPoints == 0 ) {
           throw UnitIsDead();
       }
    }

    int Unit::getDamage() const{
        return this->damage;
    }
    int Unit::getHitPoints() const{
        return this->hitPoints;
    }
    int Unit::getHitPointsLimit() const{
        return this->hitPointsLimit;
    }
    const std::string& Unit::getName() const{
        return this->name;
    }

    void Unit::addHitPoints(int hp){
        try {
        ensureIsAlive();
        int currentHitPoint = this->hitPoints;

        currentHitPoint += hp;

        if (currentHitPoint > hitPointsLimit) {
            this->hitPoints = hitPointsLimit;
        } else {
            this->hitPoints =  currentHitPoint;
        }
        } catch (UnitIsDead) {
            std::cout << "Cant Heal dead unit" << std::endl;
        }


    }
    void Unit::takeDamage(int dmg) {
        int currentHitPoint = this->hitPoints;

        currentHitPoint -= dmg;
        if (currentHitPoint <= 0) {
            currentHitPoint = 0;
        } 
        this->hitPoints = currentHitPoint;

    }

    void Unit::attack(Unit& enemy) {

        try {
        ensureIsAlive();
        enemy.takeDamage(this->damage);
        enemy.counterAttack(*this);
        } catch (UnitIsDead) {
            std::cout << "my target is dead" << std::endl;
        }
    }
    void Unit::counterAttack(Unit& enemy) {
        try {
        ensureIsAlive();
        this->takeDamage(enemy.damage/2);
        } catch (UnitIsDead) {
            std::cout << "this unit is dead" << std::endl;
        }
    }

    std::ostream& operator<<(std::ostream& out, const Unit& unit) {
        out << "damage " <<unit.getDamage() << std::endl; 
        out << "hit Points " <<unit.getHitPoints() << std::endl; 
        out << "Max HP " <<unit.getHitPointsLimit() << std::endl; 
        out << "Unit " <<unit.getName() << std::endl; 
    }